.github/scripts/check_ocp_indent.py

```diff
@@ -0,0 +1,147 @@
+    lines = tuple((await get_text(cmd=cmd, stdin=stdin, error_codes=error_codes)).splitlines())
+    
+    return lines
+
+
+@dataclass(frozen=True)
+class Hunk:
+    start: int
+    stop: int
+
+    # "git diff -U0" has the following predictable pattern for hunk headers. It shows exactly where
+    # a change starts and if any additional lines were changed after that, how many.
+    _header_pattern: Pattern = re.compile(r'^@@.*\+(?P<start>\d+),?(?P<delta>\d+)? @@.*$')
+
+    @staticmethod
+    async def from_path(path: Path, /) -> Tuple[Hunk]:
+        """Return Tuple[Hunk] of with start/stop of each hunk changed since repo's master branch."""
+
+        # For each hunk header in git diff, extract start/stop of changed lines, and create Hunk.
+        hunks: List[Hunk] = []
+        for line in await get_lines(f'git diff remotes/origin/master -U0 {path}'):
+            if match := Hunk._header_pattern.match(line):
+                start = int(match.groupdict()['start'])
+                stop = start + int(match.groupdict()['delta'] or '0')
+                hunks.append(Hunk(start, stop))
+        hunks: Tuple[Hunk] = tuple(hunks)
+        
+        return hunks
+
+
+@dataclass(frozen=True)
+class FileDiff:
+    text: Text
+
+    @staticmethod
+    async def from_path(path: Path, /) -> FileDiff:
+        """Return File with diff-text suggested by ocp-indent."""
+
+        # Update file contents by running ocp-indent on every hunk of code changed since master.
+        text = path.read_text()
+        hunks = await Hunk.from_path(path)
+        for hunk in hunks:
+            text = await get_text(
+                f'opam exec -- ocp-indent --lines={hunk.start}-{hunk.stop}', stdin=text,
+            )
+
+        # Get diff of file contents as suggested by ocp-indent and original file. Error codes 0/1
+        # indicate no/some difference respectively.
+        text = await get_text(f'diff -u {path} -', stdin=text, error_codes=frozenset({0, 1}))
+
+        return FileDiff(text=text)
+
+
+@dataclass(frozen=True)
+class CommitDiff:
+    text: Text
+
+    @staticmethod
+    async def from_env() -> CommitDiff:
+        """Return CommitDiff with diff text suggested by ocp-indent."""
+
+        # Repo path is needed to construct absolute paths for changed OCaml files.
+        repo_path = Path((await get_text('git rev-parse --show-toplevel')).strip())
+
+        # Ensure master is checked out for comparison.
+        lines = tuple(line.strip() for line in await get_lines('git branch -a'))
+        assert 'remotes/origin/master' in lines, 'local checkout of origin/master is required'
+
+        # Names of files changed since master.
+        lines = await get_lines('git diff --name-only remotes/origin/master')
+
+        # Absolute paths of all OCaml files changed since master.
+        paths = tuple(
+            path for path in [repo_path / Path(line) for line in lines]
+            if path.suffix in {'.ml', '.mli'}
+        )
```

Comment:
Wow, Python 3 is a very different beast than Python 2.  It's really cool how the generator expression works in conjunction with async/await.